File: app/src/main/java/com/billpayments/features/settings/data/SettingsRepository.kt
package com.billpayments.features.settings.data

import com.billpayments.core.storage.KeyValueStore

interface SettingsRepository {
    fun restore(): AppSettings
    suspend fun saveThemeMode(mode: AppThemeMode)
}

class StoredSettingsRepository(private val store: KeyValueStore) : SettingsRepository {

    override fun restore(): AppSettings {
        val themeName = store.read(THEME_MODE_KEY)
        val themeMode = AppThemeMode.entries.firstOrNull { it.name == themeName } ?: AppThemeMode.System
        return AppSettings(themeMode = themeMode)
    }

    override suspend fun saveThemeMode(mode: AppThemeMode) {
        store.write(THEME_MODE_KEY, mode.name)
    }

    private companion object {
        const val THEME_MODE_KEY = "settings.themeMode"
    }
}

File: BillPayments/Features/Settings/Data/SettingsRepository.swift
protocol SettingsRepository {
    func restore() -> AppSettings
    func saveThemeMode(_ mode: AppThemeMode) async throws
}

final class StoredSettingsRepository: SettingsRepository {
    private static let themeModeKey = "settings.themeMode"

    private let store: KeyValueStore

    init(store: KeyValueStore) {
        self.store = store
    }

    func restore() -> AppSettings {
        let themeName = store.read(Self.themeModeKey)
        let themeMode = AppThemeMode.allCases.first { $0.rawValue == themeName } ?? .system
        return AppSettings(themeMode: themeMode)
    }

    func saveThemeMode(_ mode: AppThemeMode) async throws {
        try await store.write(Self.themeModeKey, mode.rawValue)
    }
}
